src/libopenrave/trajectory.cpp

```diff
@@ -21,38 +21,175 @@
-    if( GetDescription().size() > 0 ) {
-        O << "<description><![CDATA[" << GetDescription() << "]]></description>" << endl;
+}
+
+inline void WriteBinaryVector(std::ostream&f, const std::vector<dReal>& v)
+{
+    // Indicate number of data points
+    const uint32_t numDataPoints = v.size();
+    WriteBinaryUInt32(f, numDataPoints);
+
+    // Write vector memory block to binary file
+    const uint64_t vectorLengthBytes = numDataPoints*sizeof(dReal);
+    f.write((const char*) &v[0], vectorLengthBytes);
+}
+
+/* Helper functions for binary trajectory file reading */
+inline bool ReadBinaryUInt16(std::istream& f, uint16_t& value)
+{
+    f.read((char*) &value, sizeof(value));
+    return !!f;
+}
+
+inline bool ReadBinaryUInt32(std::istream& f, uint32_t& value)
+{
+    f.read((char*) &value, sizeof(value));
+    return !!f;
+}
+
+inline bool ReadBinaryInt(std::istream& f, int& value)
+{
+    f.read((char*) &value, sizeof(value));
+    return !!f;
+}
+
+inline bool ReadBinaryString(std::istream& f, std::string& s)
+{
+    uint16_t length = 0;
+    ReadBinaryUInt16(f, length);
+    if (length > 0)
+    {
+        s.resize(length);
+        f.read(&s[0], length);
     }
-    if( GetReadableInterfaces().size() > 0 ) {
-        xmlreaders::StreamXMLWriterPtr writer(new xmlreaders::StreamXMLWriter("readable"));
-        FOREACHC(it, GetReadableInterfaces()) {
-            BaseXMLWriterPtr newwriter = writer->AddChild(it->first);
-            it->second->Serialize(newwriter,options);
-        }
-        writer->Serialize(O);
+    else
+    {
+        s.clear();
     }
-    O << "</trajectory>" << endl;
+    return !!f;
+}
+
+inline bool ReadBinaryVector(std::istream& f, std::vector<dReal>& v)
+{
+    // Get number of data points
+    uint32_t numDataPoints = 0;
+    ReadBinaryUInt32(f, numDataPoints);
+    v.resize(numDataPoints);
+
+    // Load binary directly to vector
+    const uint64_t vectorLengthBytes = numDataPoints*sizeof(dReal);
+    f.read((char*) &v[0], vectorLengthBytes);
+
+    return !!f;
+}
+
+// New feature: Store trajectory file in binary
+void TrajectoryBase::serialize(std::ostream& O, int options) const
+{
+    // NOTE: Ignore 'options' argument for now
+
+    // Write binary file header
+    WriteBinaryUInt16(O, MAGIC_NUMBER);
+    WriteBinaryUInt16(O, VERSION_NUMBER);
+
+    /* Store meta-data */
+
+    // Indicate size of meta data
+    const ConfigurationSpecification& spec = this->GetConfigurationSpecification();
+    const uint16_t numGroups = spec._vgroups.size();
+    WriteBinaryUInt16(O, numGroups);
+
+    FOREACHC(itgroup, spec._vgroups)
+    {
+        WriteBinaryString(O, itgroup->name);   // Writes group name
+        WriteBinaryInt(O, itgroup->offset);    // Writes offset
+        WriteBinaryInt(O, itgroup->dof);       // Writes dof
+        WriteBinaryString(O, itgroup->interpolation);  // Writes interpolation
+    }
+
+    /* Store data waypoints */
+    std::vector<dReal> trajectoryData;
```

Comment:
@ShaysRebellion please do this without creating a new vector instance (since it allocates memory)

Reply:
@rdiankov i noticed that there is no `vector<dReal>` member for waypoints in `TrajectoryBase` defined in `trajectory.h` that would allow us to read or write directly to this member (allowing us to remove the temporarily created vector). However, `GenericTrajectory` does contain such a member (called `_vtrajdata`)

i could:
1) move the entire serialization and deserialization to `GenericTrajectory` (because it seems like `GenericTrajectory` is the only class inheriting from `TrajectoryBase`, correct me if i'm wrong)
2) create a function that returns a pointer reference to the member function (declared in `trajectory.h` but implemented in `generictrajectory.cpp`
3) additional options?

Reply:
@ShaysRebellion yes, in this case, we have to override the serialize function in GenericTrajectory and do the fast thing. You can still leave the slow function in TrajectoryBase, just please put comments saying that this is a default implementation and faster implementations should be done in the specific Trajectory classes.